### ISSUE TITLE
Introduce a new resolver interface (for use by the reactor) which supports multiple addresses and multiple address families

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4362]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4362|
|Type|enhancement|
|Created|2010-03-10 22:22:16Z|
|Branch|https://github.com/twisted/twisted/tree/4362-multiresolve|

The current resolver interface, `IResolverSimple.getHostByName`, is modeled on `gethostbyname(3)` and inherits the limitations of that API:

  * Only one address can be returned
  * The desired address family cannot be specified

`IResolverSimple` tries to account for the latter by indicating that IPv6 addresses will be returned in preference to IPv4 addresses, but this leads to bugs such as [#4212](https://github.com/twisted/twisted/issues/4212), since the reactor can't actually do anything with an IPv6 address.

`getaddrinfo(3)` was introduced to address the shortcomings of `gethostbyname(3)`.  It allows multiple addresses to be returned, and includes information about the address family of each (although this may be redundant with the address itself, since an ipv4 literal and an ipv6 literal can be distinguished from each other anyway).

It also mixes in `getservbyname(3)` functionality, accepting a service name and including a port number as an attribute of each element in the result.

We should introduce an interface like `getaddrinfo(3)` and switch the reactor to prefer it.  We can register an adapter from the old interface to continue to support users of it.  The new interface will eventually provide better IPv6 functionality, though, and if used will avoid issues like [#4212](https://github.com/twisted/twisted/issues/4212).

Here's a proposal for the actual interface:

```
class INameResolver(Interface):
    def getAddressInformation(name, service, family=None, socktype=None, proto=None, flags=None):
        """
        Look up the address information associated with the given name.

        @param name: A hostname to resolve.
        @type name: C{str}

        @param service: A port number or the name of the service for which to find address
            information.  For example, C{22} or C{"ssh"}.
        @type service: C{int} or C{str}

        @param family: If specified, limit results to addresses from this family.  Must be one
            of the address family constants from the socket module.  For example,
            L{socket.AF_INET}.

        @param socktype: If specified, limit results to addresses for this socket type.  Must be
            one of the socket type constants from the socket module.  For example,
            L{socket.SOCK_STREAM}.

        @param proto: If specified, limit results to addresses for this socket protocol.  Must
            be one of the protocol constants from the socket module.  For example,
            L{socket.IPPROTO_TCP}.

        @param flags: A bitvector specifying zero or more of the following::

            - Yea right.  Go read the `getaddrinfo(3)` man page.

        @raise ValueError: If one of the specified flags is not supported by the
            implementation.  All flags are optional.

        @return: A L{Deferred} which will fire when the resolution completes.  If resolution is
            successful, the result will be a list of objects with the following attributes:

            - C{family}: the family of this address
            - C{type}: the type of this address
            - C{protocol}: the protocol of this address
            - C{canonicalName}: the canonical name associated with this address, or an empty
                                string.
            - C{address}: The actual address itself, including a port number.  This is suitable
                          to be passed directly to L{socket.socket.connect}.
        """```{
class INameResolver(Interface):
    def getAddressInformation(name, service, family=None, socktype=None, proto=None, flags=None):
        """
        Look up the address information associated with the given name.

        @param name: A hostname to resolve.
        @type name: C{str}

        @param service: A port number or the name of the service for which to find address
            information.  For example, C{22} or C{"ssh"}.
        @type service: C{int} or C{str}

        @param family: If specified, limit results to addresses from this family.  Must be one
            of the address family constants from the socket module.  For example,
            L{socket.AF_INET}.

        @param socktype: If specified, limit results to addresses for this socket type.  Must be
            one of the socket type constants from the socket module.  For example,
            L{socket.SOCK_STREAM}.

        @param proto: If specified, limit results to addresses for this socket protocol.  Must
            be one of the protocol constants from the socket module.  For example,
            L{socket.IPPROTO_TCP}.

        @param flags: A bitvector specifying zero or more of the following::

            - Yea right.  Go read the `getaddrinfo(3)` man page.

        @raise ValueError: If one of the specified flags is not supported by the
            implementation.  All flags are optional.

        @return: A L{Deferred} which will fire when the resolution completes.  If resolution is
            successful, the result will be a list of objects with the following attributes:

            - C{family}: the family of this address
            - C{type}: the type of this address
            - C{protocol}: the protocol of this address
            - C{canonicalName}: the canonical name associated with this address, or an empty
                                string.
            - C{address}: The actual address itself, including a port number.  This is suitable
                          to be passed directly to L{socket.socket.connect}.
        """


Attachments:

* [getaddrinfo-4362-3-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/384/38481c57a34577b7a292e38454908470990741cf/c72258cff383fb460952dca58622a337bf3aeb52.patch) (14751 bytes) - added by rwall on 2013-02-11 14:56:32Z - Merged forward and resolved conflicts with trunk - contains broken tests

<details><summary>Searchable metadata</summary>

```
trac-id__4362 4362
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__4362_multiresolve 4362-multiresolve
branch_author__exarkun__glyph__rwall exarkun, glyph, rwall
status__closed closed
resolution__fixed fixed
component__core core
keywords__ipv6 ipv6
time__1268259736000000 1268259736000000
changetime__1480146705975795 1480146705975795
version__None None
owner__glyph glyph
cc__jknight cc__twm cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [28686]) Branching to 'getaddrinfo-4362'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Got a reasonable start on this.  Some things I know are left to do:

  1. Finish testing all the cases of `_ResolverComplexifier`
  1. Add something like `ThreadedResolver` for `INameResolver`
  1. Make reactor.resolver be an `IResolverSimple` again
  1. Add an adapter from `INameResolver` to `IResolverSimple` to preserve backwards compatibility of reactor.resolver
  1. Think about a new reactor name resolution interface (and think about not providing one).
  1. Deprecate `IReactorCore.resolve`
  1. Deprecate reactor.resolver

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

[#4912](https://github.com/twisted/twisted/issues/4912) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

I think a `Deferred` is the wrong type for this new API to return.

One of the deficiencies of `getaddrinfo` is that it blocks.  Obviously a `Deferred` resolves that.  But another deficiency of `getaddrinfo` is that it doesn't complete (blocking or not) until all of its results are ready; certain types of requests (for example: mdns) may take significantly longer than others, and it would be great to get a successful result quickly if that is possible.  Platform-specific interfaces such as [getaddrinfo_a](http://www.kernel.org/doc/man-pages/online/pages/man3/getaddrinfo_a.3.html) and [CFHostStartInfoResolution](https://developer.apple.com/library/mac/#documentation/CoreFoundation/Reference/CFHostRef/Reference/reference.html) can invoke callbacks for each name-resolution attempt, whether it succeeded or failed.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_7"></a>|@jyknight commented|
|-|-|

I disagree. I don't think it makes sense or is useful to return partial results for one name look up. Getaddrinfo_a doesn't. It takes a *list* of hostnames to resolve. This API doesn't and shouldn't do that, so a Deferred is sufficient.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Replying to [jknight](#note_7):
> I disagree. I don't think it makes sense or is useful to return partial results for one name look up. Getaddrinfo_a doesn't. It takes a *list* of hostnames to resolve. This API doesn't and shouldn't do that, so a Deferred is sufficient. 

OK.  Thanks for pointing out that detail; looking at the definition of `struct gaicb` now it is obvious, of course.  I guess a `Deferred` that fires a `list` is the right structure after all.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Asynchronous stuff is hard in C.  It's easier in Twisted, I hope.  "glibc _can't_ do incremental asynchronous resolution" sounds like a bad justification for "Twisted _shouldn't_ do incremental asynchronous resolution".

Maybe there's another reason, though?

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_11"></a>|@wallrj commented|
|-|-|

(In [37217]) Branching to 'getaddrinfo-4362-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_12"></a>|@wallrj commented|
|-|-|

(In [37218]) merge forward by applying getaddrinfo-4362-3-1.patch. refs [#4362](https://github.com/twisted/twisted/issues/4362)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_13"></a>|@wallrj commented|
|-|-|

[Forced build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/getaddrinfo-4362-2)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj commented|
|-|-|

(In [37251]) Made test_failure pass. Instead of passing an explicit failure to
FakeResolver, just request an unknown name.

Also some smaller changes...
 * Various small docstring improvements
 * Fix a  twistedchecker comment init caps warning.
 * Add doc strings for FakeResolver
 * Fix long line
 * Add docstrings for MemoryNameResolver
 * More complete docstrings for _ResolverComplexifier
 * Add docstrings for test_base.py
 * Add docstrings for AddressInformation
 * Added a placeholder ThreadedNameResolver to make the tests happier.

refs [#4362](https://github.com/twisted/twisted/issues/4362)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Hi exarkun,

I've done a bit more work on this ticket but now I need some
guidance.

First here's what I've done on top of you original branch (in
log:branches/getaddrinfo-4362-2)
 1. Merged forward resolving some conflicts
 2. Added ThreadedNameResolver + tests
 3. Added a timeout parameter to INameResolver.getAddressInformation
 4. Registered an IResolverSimple to INameResolver adapter.

I think that covers points 1 and 2 in your comment:2 (above)

I have some questions about your other points.

> Make reactor.resolver be an IResolverSimple again

Is this for backwards compatibility?
So instead of adapting the resolver in reactor.installResolver...
we would do the adaption inside reactor.resolve everytime it's
called.
or not at all.

> Add an adapter from INameResolver to IResolverSimple to preserve
>  backwards compatibility of reactor.resolver

So maybe you meant
```

def installResolver(self, resolver):
    self._resolver = INameResolver(resolver)

```

reactor.resolver then becomes a descriptor which issues a deprecation
warning and returns IResolverSimple(reactor._resolver)

> Think about a new reactor name resolution interface (and think about
>  not providing one).

So reactor.resolve must continue to return an IP address (as per IReactorCore)

In which case why bother changing any of the reactor code in this
branch.

Why can't this branch simply be about adding INameResolver,
ThreadedNameReactor, and optionally the deprecations below...

Then decide on the reactor interface based on the usage of getaddrinfo
in [#4859](https://github.com/twisted/twisted/issues/4859).

 * Deprecate IReactorCore.resolve
 * Deprecate reactor.resolver

At which point I think I understand why you paused work on this branch
in the first place :)

Anyway. It's been interesting to learn a little more about the gai
etc.

If you have any time to comment I'd be interested hear what you think.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

Replying to [rwall](#note_15):
> Hi exarkun,
> 
> I've done a bit more work on this ticket but now I need some
> guidance.
> 
> I have some questions about your other points.
> If you have any time to comment I'd be interested hear what you think.
> 
> -RichardW.

Assigning it to exarkun for rwall's questions.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph set owner to @wallrj|
|-|-|

Replying to [rwall](#note_15):
> Hi exarkun,

I'll be filling in for exarkun today.

> I've done a bit more work on this ticket but now I need some
> guidance.

Thanks!

> First here's what I've done on top of you original branch (in
> log:branches/getaddrinfo-4362-2)
>  1. Merged forward resolving some conflicts
>  2. Added ThreadedNameResolver + tests
>  3. Added a timeout parameter to INameResolver.getAddressInformation
>  4. Registered an IResolverSimple to INameResolver adapter.

Great.

> I think that covers points 1 and 2 in your comment:2 (above)
> 
> I have some questions about your other points.
> 
> > Make reactor.resolver be an IResolverSimple again
> 
> Is this for backwards compatibility?

Yes.  The backwards compatibility contract here is that the `resolver` attribute is an `IResolverSimple`.

> So instead of adapting the resolver in reactor.installResolver...
> we would do the adaption inside reactor.resolve everytime it's
> called.
> or not at all.

Well... you don't need to do that.  Just because `resolver` is a public attribute does not mean you need to consult it each time you do a resolution.  You can do the adaptation in `installResolver` as long as the new attribute you're storing is under a different name.  You may even want to make the new attribute public, if there are reasons that callers would want to access it directly.

> > Add an adapter from INameResolver to IResolverSimple to preserve
> >  backwards compatibility of reactor.resolver
> 
> So maybe you meant
> ...
> reactor.resolver then becomes a descriptor which issues a deprecation
> warning and returns IResolverSimple(reactor._resolver)

Right, exactly.

> > Think about a new reactor name resolution interface (and think about
> >  not providing one).
> 
> So reactor.resolve must continue to return an IP address (as per IReactorCore)

Yup, them's the breaks.  It has a contract, it can't change the contract.

> In which case why bother changing any of the reactor code in this branch.
> 
> Why can't this branch simply be about adding INameResolver,
> ThreadedNameReactor, and optionally the deprecations below...

I'm definitely in favor of smaller branches.

> Then decide on the reactor interface based on the usage of getaddrinfo
> in [#4859](https://github.com/twisted/twisted/issues/4859).
> 
>  * Deprecate IReactorCore.resolve
>  * Deprecate reactor.resolver
> 
> At which point I think I understand why you paused work on this branch
> in the first place :)
> 
> Anyway. It's been interesting to learn a little more about the gai
> etc.
> 
> If you have any time to comment I'd be interested hear what you think.

Did that answer most of your questions about this?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Today I discovered an interesting misfeature: since we don't have this reactor interface and `HostnameEndpoint` directly spins off `getaddrinfo` into a thread using `deferToThread`, `HostnameEndpoint` becomes somewhat test-hostile.  Which, because `pyOpenSSL` is _also_ test-hostile, [#5642](https://github.com/twisted/twisted/issues/5642) is hard to implement nice expressive tests for.  So I am going to bump the priority and do some work on this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

We managed to test [#5642](https://github.com/twisted/twisted/issues/5642), but the absence of this interface makes every test of `HostnameEndpoint` overly fragile and potentially buggy; there's a private `makeHostnameEndpointSynchronous` helper which is not available to applications.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set owner to @glyph|
|-|-|

This is now effectively blocking [#6712](https://github.com/twisted/twisted/issues/6712) so I will try to pick it up this week.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

(In [46789]) Branching to getaddrinfo-4362-3.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph commented|
|-|-|

We need a github watcher that pushes this 'branch' field update :-\

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_24"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_9):
> Asynchronous stuff is hard in C.  It's easier in Twisted, I hope.  "glibc _can't_ do incremental asynchronous resolution" sounds like a bad justification for "Twisted _shouldn't_ do incremental asynchronous resolution".
> 
> Maybe there's another reason, though?

Looks like this question never got answered, and since I'm in the process of retooling the interface, it should probably be addressed.  And in the intervening 4 years, maybe my opinion has changed a little.

Happy Eyeballs repeatedly references the "host's address preference policy", which can apparently be configured via DHCPv6, as per this citation: https://tools.ietf.org/html/rfc6555#ref-ADDR-SELECT.  So with `getaddrinfo` and the like (all the wacky platform-specific variants seem the same, as James noted with `getaddrinfo_a` and I've been able to discover on Windows, e.g. https://msdn.microsoft.com/en-us/library/windows/desktop/ms738518(v=vs.85).aspx), all the hostname resolutions are going to come back in one big bundle, because they need to be sorted after the last one arrives.

However.

While the spec says that `getaddrinfo` should behave that way, that does _not_ mean that the baseline interface for doing hostname resolution _must_ behave this way.  In a high-scale web spider, for example, the "host address preference policy" could be considered to be encapsulated in the application's twisted.names configuration, and that policy could be "resolve everything as fast as possible".

This might even be useful when calling `getaddrinfo` itself.  It's definitely not unheard of for some networks to break horribly when you try to do IPv6 resolution but behave sensibly for IPv4.  We could cheat the specs a little bit and kick off a parallel v4-only resolution if the v6 resolution is taking an appreciable amount of time.  However, it would be impossible to relay all the addresses back to the application promptly if we were restrited by a `getaddrinfo` style all-at-once return API.

So, after much consideration I am thinking we _do_ want to have an interface which gives back name resolutions incrementally, to make that possible in the cases where it is possible and we care.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph commented|
|-|-|

I updated the branch here - https://github.com/twisted/twisted/pull/547 - and I was planning to try to put it back into review today, but after spending hours re-understanding the problem domain here I think i'm going to start this branch over from scratch with a streaming interface, one which presents something in terms of an `IProtocol`-style object that receives `resolutionStarted`, `addressReceived`, and `resolutionComplete` events; the addresses in question should be `IPv4Address` and `IPv6Address` objects.  `HostnameEndpoint` will have to change to manage a little state machine that cancels an outstanding name resolution if one is in progress when it successfully connects.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_26"></a>|@glyph commented|
|-|-|

New PR up, definitely not ready for review, here: https://github.com/twisted/twisted/pull/548

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph commented|
|-|-|

A couple of notes on the new proposed interface in that branch:

1. it doesn't expose port-number resolution. this is a dumb thing. nobody types service names. if they did, the wording of various specs constrain the platform's behavior to never doing anything but looking in `/etc/services`.  Apparently honoring `getaddrinfo` can even get you in trouble here, since some OSes (iOS 9, maybe others) sometimes have bugs where they forget to fill out the 'port' field of the address.  the only utility of the port argument to getaddrinfo is to pre-populate a data structure, not to do any meaningful resolution.  so: pass an integer, just have it populate, skip the dependence on whatever random data happens to be in one machine's `/etc/services` file.
2. it doesn't expose the canonical name anywhere.  this seems like losing potentially useful information, except, that information can only really be used for one purpose: to screw up which TLS name to actually validate against by trusting local DNS.  Don't do it!
3. it carefully avoids talking about `AF_INET`,  `SOCK_STREAM`, `PF_WHATEVER` and `IPPROTO_NONSENSE`.  The type language we've invented in Twisted for address families isn't the absolute best, and hopefully eventually we can overhaul it, but it's already there, and it's definitely worse to make users have to understand _two_ crappy mini-langauges about transport protocols.

I did add scope ID and flow info to `IPv6Address`, though, because those are kinda important for some edge-cases of IPv6 connectivity (particularly scope ID).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_28"></a>|@glyph commented|
|-|-|

Since I've gotten no feedback, I've started implementing a little bit.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_30"></a>|@glyph removed owner|
|-|-|

OK. Please review this PR: https://github.com/twisted/twisted/pull/548

It's forward and backward compatible, it handles multiple address families and multiple address types, and it uses the existing address objects within Twisted to communicate all this stuff.

Still to do beyond the end of this branch:

- make `HostnameEndpoint` actually take advantage of the newly asynchronous interface on the reactor, (being sure to check that it's available first, for the case of 3rd-party reactors!)
- expose some of the deterministic threadpool testing stuff here more centrally within Twisted
- handle more error cases from application code
- implement `twisted.names`-based resolvers

However, it's already quite large (+1000 lines, although many of them are quite spacious docstrings), and I didn't want it to get any bigger before landing something.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @glyph|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_32"></a>|@glyph set status to `closed`|
|-|-|

In changeset 5178c338caa95fd5ded1f4b0455fb75e22d99ab7
```
#!CommitTicketReference repository="" revision="5178c338caa95fd5ded1f4b0455fb75e22d99ab7"
Merge 4362-multiresolve: add IHostnameResolver

Author: glyph

Reviewer: hawkowl

Fixes: #4362

Added a new interface, twisted.internet.interfaces.IHostnameResolver, which is
an improvement to twisted.internet.interfaces.IResolverSimple that supports
resolving multiple addresses as well as resolving IPv6 addresses.  This is a
native, asynchronous, Twisted analogue to getaddrinfo.
```
